### ISSUE TITLE
Memory leak testing

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/calculus-grapher#265]
For https://github.com/phetsims/calculus-grapher/issues/262.

We can do this at the same time as code review, but should complete it before QA testing begins (3/20).

We'll need to do this for both phet and phet-io brands.

**Comment by pixelzoom** [comment]
Here's the `brand=phet` memory test.

* Build = `grunt dev --minify.mangle=false --brands=phet,phet-io`
* URL = https://phet-dev.colorado.edu/html/calculus-grapher/1.0.0-dev.22/phet/calculus-grapher_en_phet.html?fuzz

Snaphot 1 was taken immediately after starting the sim.  Snaphots 2-11 were taken at 1-minute intervals.  Snapshot 12 was taken 10 minutes after Snapshot 11.   

The memory footprint is very large -- perhaps that's due to `--minify.mangle=false`.  More importantly, the sim appears to have a decent memory leak.  Heap size decreased very slightly (1MB) from Snapshot 7 to 8.   But it otherwise continued to increase, starting at 64MB and finishing at 142MB.  I have not investigated the cause.

<img width="226" alt="screenshot_2395" src="https://user-images.githubusercontent.com/3046552/223879213-53be1c09-11ae-4f49-85bc-7c9781dd38ba.png">

**Comment by pixelzoom** [comment]
Results of `brand=phet` memory test ...

* Build = `grunt dev --minify.mangle=false --brands=phet,phet-io`
* URL = https://phet-dev.colorado.edu/html/calculus-grapher/1.0.0-dev.22/phet-io/calculus-grapher_all_phet-io.html?postMessageOnError&phetioStandalone&fuzz

Snaphot 1 was taken immediately after starting the sim.  Snaphots 2-11 were taken at 1-minute intervals.  Snapshot 12 was taken 10 minutes after Snapshot 11.   

I'm very confused.  Unlike the `brand=phet` test, this test shows no sign of a memory leak.  Heap size seems to stabilize around 82MB and goes down 3 times.

Maybe the `brand=phet` test was different because I ran the build from my local calculus-grapher/build/phet/ directory, instead of from phet-colorado.edu? That makes no sense. But I'm going to re-run `brand=phet` from phet.colorado.edu.

<img width="234" alt="screenshot_2396" src="https://user-images.githubusercontent.com/3046552/223883322-4b58b2b1-feeb-43ed-a41b-dc5890e20ccf.png">

**Comment by pixelzoom** [comment]
@veillette pointer out that the above tokens appear in PhET repo fenster.  It looks like 
[@jessegreenberg](https://phetsims.slack.com/team/U6ELY17N3)
 created most of the files in the fenster repo, whose purpose (according to README.md) is to test haptics using Cordova.   Why is this being pulled into Calculus Grapher? It has no support for haptics.

**Comment by pixelzoom** [comment]
I tried to reproduce in unbuilt geometric-optics, but could not.  Heap size started at 54MB, then stabilized around 80MB for 10 heap snapshots.  So I built a local version of geometric-optics with `grunt --minify.mangle=false`. The heap size when the sim starts is consistently 124MB !!  geometric-optics does include a11y features. Has something really large (fenster? Cordova?) been added that’s getting loaded.   This sim had a memory test recently by me and QA, and it was not behaving like this.

I’m going to pause work on this issue until I hear from @jessegreenberg or someone from the a11y team.

**Comment by samreid** [comment]
I see those things defined in he-1.1.1.js, it’s a sherpa dependency: https://github.com/mathiasbynens/he

**Comment by jessegreenberg** [comment]
fenster is not a sim dependency so it cannot be related to this. I am not aware of any new a11y features or anything that would add so much to the heap size. I did a little bit of memory testing and saw some odd things:

<details><summary>Without fuzzing, just changing screens and pressing some buttons.</summary>

![image](https://user-images.githubusercontent.com/6396244/223915016-d46e0b68-f3d1-487b-863f-a1513d6f5b88.png)
</details>



<details><summary>With fuzzing, snapshot every 30 seconds</summary>

![image](https://user-images.githubusercontent.com/6396244/223916107-94e179db-aa64-4643-8992-e39577bb2c5e.png)

</details>

I think `he` is used by RichText, is anything creating and never disposing a lot of RichText? Also, does this mean that there are 102 LinePlots in the sim?
![image](https://user-images.githubusercontent.com/6396244/223916884-d94fd8e6-7ecc-4698-9007-d9c4c1e017ce.png)

**Comment by jessegreenberg** [comment]
OK, thanks. Maybe LinePlot just generates a ton of garbage when it updates. I think(?) this readout shows us the heap every second or so without pressing "Take Snapshot".

![image](https://user-images.githubusercontent.com/6396244/224063539-6417f5e3-e5cc-4bec-abf8-3a9ef5a4dc3c.png)

Here are some values I am seeing:
- On load: 185 MB
- Press Collect Garbage button: 87.7 MB
- Move the graph a bit on the "Derivative" screen: 203 MB
- Press Collect Garbage button: 88.2 MB

and so on, this is pretty reproducible for me.

**Comment by pixelzoom** [comment]
Thanks @jessegreenberg. And sorry to drag you into this.

**Comment by pixelzoom** [comment]
While working on Undo in #273, it occurred to me that the Undo feature might be responsible for the sudden memory spikes. Each time a curve is changed, 1251 instance of:

```
type CurvePointState = {
  y: number;
  pointType: PointType;
}
```

... are pushed onto the Undo stack.  The maximum stack length is 20, so there are 20 x 1251 = 25020 `CurvePointState` objects that can be very suddenly created by a fuzzer that is rapidly "clicking" in the graphs.

We could test this hypothesis by commenting out the body of CurvePoint `save`, so that it's a no-op. Then test to see if the memory spikes are still occurring.

**Comment by veillette** [comment]
hhmm interesting, that should be an easy check:
FYI: `CalculusGrapherConstants` is hoisting the constant 
``
// maximum of undo actions (See https://github.com/phetsims/calculus-grapher/issues/64)
  MAX_UNDO: 20,
``

**Comment by veillette** [comment]
I'll test your hypothesis. I will not fuzz such that I have maximum control. I'll test the predict curve, which is also saved, but does not generate additional noise by not having derivative curves, etc, attached  to it. We can add 20 curves, take a snapshot, then undo them, as see if we come back to our starting memory footprint.

**Comment by veillette** [comment]
Tested with 
http://localhost:63342/phetsims/calculus-grapher/calculus-grapher_en.html?brand=phet&ea&debugger&screens=1

I start only with screen 1, so therefore I start with a smaller footprint. I turned on the predict mode.

Snapshot 1 is the baseline
snapshot 2: I applied a curve 10 times
snapshot 3: I applied a curve 10 times, so 20 total
snapshot 4: I applied a curve 10 times, so 30 total. Notice that the memory should not increase, and it doesnt.
snapshot 5: I undo the curve 10 times.
snapshot 6: I undo the curve 10 times.  The stack is now empty.

I should also note that right before I collected a snapshot, I did a manual garbage collection to get rid of any noise.

<img src="https://user-images.githubusercontent.com/7633344/224342619-f06555e1-c09a-4139-9c0d-c890bbcacac9.png" width=200>

The bottom line, is that the stack has minimum impact on the memory, of the order of 1 MB

**Comment by pixelzoom** [comment]
Great, thanks for testing the Undo feature.

**Comment by pixelzoom** [comment]
I'm seeing the same thing that @jessegreenberg observed in https://github.com/phetsims/calculus-grapher/issues/265#issuecomment-1462204007.   Memory footprint varies wildly as the sim is exercised. And I can identify anything that's clearly a memory leak.

Can we safely conclude that  this is to be expected with so many bamboo plots, with large dataSets?  And just alert QA to this fact before testing? @veillette what's your opinion?

**Comment by veillette** [comment]
Memory test on Master 

Browser: Chrome 110.0.5481.180 (
URL: http://localhost:63342/phetsims/calculus-grapher/calculus-grapher_en.html?brand=phet&ea&debugger&fuzz

Test protocol: Same as above, plus:

All Snapshots are taken at 2 minutes intervals 
There is a 10 minutes interval between snapshot 9 and 10.

The memory increased for all snapshots.


![image](https://user-images.githubusercontent.com/7633344/224773532-e82df853-1849-490b-ba1c-18dee2b2c144.png)


Here is the difference between snapshot 9 and 2.

<html>
<body>


(number) | 844 023 | 719 638 | +124 385 | 10 128 276 | 8 635 656 | +1 492 620 |  
-- | -- | -- | -- | -- | -- | -- | --
Object | 215 586 | 140 666 | +74 920 | 4 658 240 | 3 113 816 | +1 544 424 |  
Array | 380 701 | 317 360 | +63 341 | 6 091 264 | 5 077 808 | +1 013 456 |  
(array) | 314 135 | 260 772 | +53 363 | 24 525 208 | 20 135 888 | +4 389 320 |  
Set | 217 188 | 179 943 | +37 245 | 3 475 056 | 2 879 136 | +595 920 |  
(closure) | 171 432 | 150 712 | +20 720 | 4 356 716 | 3 845 592 | +511 124 |  
TinyEmitter | 140 751 | 120 142 | +20 609 | 3 378 000 | 2 883 384 | +494 616 |  
Array | 380 464 | 368 856 | +11 608 | 6 087 472 | 5 901 744 | +185 728 |  
Bounds2 | 80 648 | 70 668 | +9 980 | 4 193 696 | 3 674 736 | +518 960 |  
(number) | 841 455 | 831 849 | +9 606 | 10 097 460 | 9 982 188 | +115 272 |  
(array) | 313 387 | 304 454 | +8 933 | 24 471 820 | 23 665 152 | +806 668 |  
Set | 215 815 | 209 618 | +6 197 | 3 453 076 | 3 353 924 | +99 152 |  
TinyForwardingProperty | 22 524 | 16 358 | +6 166 | 900 960 | 654 320 | +246 640 |  
TinyProperty | 36 210 | 30 369 | +5 841 | 1 158 600 | 971 688 | +186 912 |  
Object | 215 286 | 209 671 | +5 615 | 4 652 224 | 4 531 968 | +120 256 |  
system / Context | 44 543 | 39 698 | +4 845 | 950 864 | 853 176 | +97 688 |  
Matrix3 | 19 905 | 15 214 | +4 691 | 477 720 | 365 136 | +112 584 |  
TinyStaticProperty | 17 220 | 12 612 | +4 608 | 619 920 | 454 032 | +165 888 |  
(closure) | 171 001 | 166 538 | +4 463 | 4 345 748 | 4 237 020 | +108 728 |  
(compiled code) | 97 870 | 94 329 | +3 541 | 12 610 156 | 10 902 600 | +1 707 556 |  
TinyEmitter | 139 347 | 135 895 | +3 452 | 3 344 316 | 3 261 468 | +82 848 |  
Vector2 | 134 033 | 131 490 | +2 543 | 2 680 652 | 2 629 792 | +50 860 |  
PDOMDisplaysInfo | 4 305 | 3 153 | +1 152 | 86 100 | 63 060 | +23 040 |  
Picker | 4 305 | 3 153 | +1 152 | 292 740 | 214 404 | +78 3


</body>
</html>

**Comment by veillette** [comment]
I don't know what to think. Is it acceptable? It is a rather significant heap increase.

**Comment by pixelzoom** [comment]
I'm seeing the same thing - memory just keeps increasing, so we probably do have a memory leak.  And with Heap Comparision, I feel like I'm looking for a needle in haystack.   I think we need to request some outside help here.

**Comment by samreid** [comment]
Notes to self for testing:

Building with:
`grunt --brands=phet --minify.mangle=false`
5 minute warm up fuzz
Use `phet.chipper.queryParameters.fuzz = !phet.chipper.queryParameters.fuzz;` and reset all between snapshots

**Comment by samreid** [comment]
It does look like a leak, can you please check usages of the CalculusGrapherSymbols including but not limited to `fStringProperty` to see their usage sites, and make sure to dispose/remove listeners?  @veillette or @pixelzoom want to schedule synchronous collaboration?

<img width="1429" alt="image" src="https://user-images.githubusercontent.com/679486/225096584-690f3854-20a0-46f1-aa21-3fd4ed7ff660.png">

**Comment by pixelzoom** [comment]
CalculusGrapherSymbols fields are used in these 3 places:

* GraphTypeLabelNode - the labels that in the upper-left corner of graphs, on radio buttons, and on the "Notation" and "Variable" preferences controls
* ReferenceLineNode - the view for the thing that appears when you click "Reference Line" checkbox
* VariableControl (the "Variable" control in the Preferences dialog) 

So perhaps this is related to disposal of the Preferences dialog. My understanding is that it's instantiate on demand, and then reused.  And in PhET-iO sims, instantiated to create the archetype.  So `dispose` methods in the various subcomponents of CalculusGrapherSimulationPreferencesNode need some work.

But how could this be accounting for such a large memory leak? Or am I mistaken about the Preference dialog being reused?

**Comment by pixelzoom** [comment]
> My understanding is that the Preferences dialog is instantiated on demand, and then reused. And in PhET-iO sims, it is also instantiated to create the archetype. 

I verified that this is indeed what is happening in phet and phet-io brands.

**Comment by samreid** [comment]
It seems like the fStringProperty.tinyProperty has 979 listeners.  Is that to be expected?

<img width="1085" alt="image" src="https://user-images.githubusercontent.com/679486/225112864-e8c03a3b-7ed8-46ca-8fbc-c353273375a2.png">

**Comment by pixelzoom** [comment]
No, it makes no sense that `fStringProperty` has 979 listeners. 

It's a dependency of 4 DerivedProperties in GraphTypeLabelNode, which is in turn used as a label in these places:

* 4 _f(x)_ radio button buttons, one for each screen (turn on 'Predict' preference to see this radio button)
* 2 graphs labels in Derivative screen
* 2 graph labels in Integral screen
* 3 graph labels and 2 radio button labels in the Advanced screen
* 4 graph labels and 2 radio button labels in the Lab screen
* 2 radio buttons for the Notation preference control

So there should be on the order of 17 listeners.  I have no explanation for why there are 979 listeners.

**Comment by pixelzoom** [comment]
Over in https://github.com/phetsims/calculus-grapher/issues/289, I address some minor leakage that was coming from `preferencesDialogCapsule.archetype`.

**Comment by pixelzoom** [comment]
I added this to CalculusGrapherScreenView `step`:

```
    console.log( `CalculusGrapherStrings.symbol.fStringProperty ${CalculusGrapherStrings.symbol.fStringProperty.getListenerCount()}` );
    console.log( `CalculusGrapherSymbols.fStringProperty ${CalculusGrapherSymbols.fStringProperty.getListenerCount()}` );
```

After several minutes of fuzzing, the maximum listener counts that I see are:

> CalculusGrapherStrings.symbol.fStringProperty 4
> CalculusGrapherSymbols.fStringProperty 28

Nothing even approaching 979.    So I'm not sure what https://github.com/phetsims/calculus-grapher/issues/265#issuecomment-1468688654 is telling us.

**Comment by samreid** [comment]
I ran the following test:

* Local build via `grunt --brands=phet --minify.mangle=false --lint=false`
* Launch built version
* Fuzz for about 30 minutes. Pause fuzzing via `phet.chipper.queryParameters.fuzz = !phet.chipper.queryParameters.fuzz;` Reset each screen, then go to screen 1
* Memory snapshot
* Fuzz for 5 minutes. Pause fuzzing via `phet.chipper.queryParameters.fuzz = !phet.chipper.queryParameters.fuzz;`. Reset each screen then go to screen 1.

Between these snapshots, I see about the same memory:
<img width="144" alt="image" src="https://user-images.githubusercontent.com/679486/225166171-8212fe0d-eeb5-4048-bcdf-1399385007df.png">

And the memory deltas look like noise:
<img width="1062" alt="image" src="https://user-images.githubusercontent.com/679486/225166243-daa4788b-df08-4221-a5b5-c69009e72787.png">

So I'm wondering if this was corrected in https://github.com/phetsims/calculus-grapher/issues/289.  Self-unassigning.

**Comment by pixelzoom** [comment]
> ... So I'm wondering if this was corrected in https://github.com/phetsims/calculus-grapher/issues/289.

Possible, I guess.  But I'm surprised that a leak in the Preferences dialog would result in such a dramatic memory spike.

Thanks for assisting @samreid, we'll take it from here.

**Comment by pixelzoom** [comment]
Results of memory test, after fixing leaks in https://github.com/phetsims/calculus-grapher/issues/289.

Device: MacBookPro16,1 (2019 Intel), macOS 13.21.
Browser: Chrome 111.0.5563.64
Build: `grunt dev --minify.mangle=false --brands=phet,phet-io`
URL: https://phet-dev.colorado.edu/html/calculus-grapher/1.0.0-dev.24/phet-io/calculus-grapher_all_phet-io.html?fuzz

Protocol: 
* Snaphot 1 was taken immediately after starting the sim.
* Snaphots 2-11 were taken at 1-minute intervals.
* Snapshot 12 was taken 5 minutes after Snapshot 11.

Results for `brand=phet`: Memory stabilizes around 81MB, then goes up and down. My conclusion is that our memory leak is resolved.

<img width="224" alt="screenshot_2437" src="https://user-images.githubusercontent.com/3046552/225655359-306ec7cc-66c9-48d8-a4c1-152731ef4d46.png">

**Comment by pixelzoom** [comment]
I'll do a test for `brand=phet-io` with 1.0.0-dev.24 later today.  If that looks OK, we can close this issue.

**Comment by pixelzoom** [comment]
Below are the results for `brand=phet-io`, with same details are https://github.com/phetsims/calculus-grapher/issues/265#issuecomment-1472131813.

Snapshot 2 is an outlier; I can rationalize this as a snapshot where GC had not yet occurred.  Otherwise, memory again stabilizes ~81MB, then goes up and down.  So our memory leak is also resolved for PhET-iO.

<img width="168" alt="screenshot_2439" src="https://user-images.githubusercontent.com/3046552/225783786-9cb982cf-361d-4541-9f2a-0549716eba8a.png">

**Comment by pixelzoom** [comment]
Conclusion: https://github.com/phetsims/calculus-grapher/issues/289 resolved the memory leak.

This issue is ready to close.
